RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronSeedProducer.cc

```diff
@@ -0,0 +1,529 @@
+
+//////////////////////////////////////////////////////////////////////////////////////////
+//
+void LowPtGsfElectronSeedProducer::beginLuminosityBlock( edm::LuminosityBlock const&, 
+							 edm::EventSetup const& setup ) 
+{
+  setup.get<IdealMagneticFieldRecord>().get(field_);
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+//
+void LowPtGsfElectronSeedProducer::produce( edm::Event& event, 
+					    const edm::EventSetup& setup ) 
+{
+  
+  // Products
+  auto seeds = std::make_unique<reco::ElectronSeedCollection>();
+  auto ecalPreIds = std::make_unique<reco::PreIdCollection>();
+  auto hcalPreIds = std::make_unique<reco::PreIdCollection>();
+  
+  // KF tracks
+  edm::Handle<reco::TrackCollection> kfTracks;
+  if ( !usePfTracks_ ) { event.getByToken(kfTracks_, kfTracks); }
+
+  // PF tracks
+  edm::Handle<reco::PFRecTrackCollection> pfTracks;
+  if ( usePfTracks_ ) { event.getByToken(pfTracks_, pfTracks); }
+
+  // ECAL clusters
+  edm::Handle<reco::PFClusterCollection> ecalClusters;
+  event.getByToken(ecalClusters_,ecalClusters);
+
+  // HCAL clusters (only used with PF tracks)
+  edm::Handle<reco::PFClusterCollection> hcalClusters;
+  event.getByToken(hcalClusters_,hcalClusters);
+
+  if ( usePfTracks_ ) { 
+    loop(pfTracks, // PF tracks
+	 ecalClusters,
+	 hcalClusters,
+	 *seeds,
+	 *ecalPreIds,
+	 *hcalPreIds,
+	 event,
+	 setup);
+  } else { 
+    loop(kfTracks, // KF tracks
+	 ecalClusters,
+	 hcalClusters,
+	 *seeds,
+	 *ecalPreIds,
+	 *hcalPreIds,
+	 event,
+	 setup);
+  }
+
+  event.put(std::move(seeds));
+  event.put(std::move(ecalPreIds));
+  event.put(std::move(hcalPreIds),"HCAL");
+  
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+// Return reco::Track from edm::Ref<T>
+
+reco::TrackRef LowPtGsfElectronSeedProducer::getBaseRef( edm::Handle< std::vector<reco::Track> > handle, int idx ) const 
+{
+  return reco::TrackRef(handle,idx);
+}
+
+reco::TrackRef LowPtGsfElectronSeedProducer::getBaseRef( edm::Handle< std::vector<reco::PFRecTrack> > handle, int idx ) const
+{
+  return reco::PFRecTrackRef(handle,idx)->trackRef();
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+// Template function, instantiated for both reco::Tracks and reco::PFRecTracks 
+template <typename T>
+void LowPtGsfElectronSeedProducer::loop( const edm::Handle< std::vector<T> >& handle, // PF or KF tracks
+					 edm::Handle<reco::PFClusterCollection>& ecalClusters,
+					 edm::Handle<reco::PFClusterCollection>& hcalClusters,
+					 reco::ElectronSeedCollection& seeds,
+					 reco::PreIdCollection& ecalPreIds, 
+					 reco::PreIdCollection& hcalPreIds,
+					 edm::Event& event,
+					 const edm::EventSetup& setup )
+{
+  
+  // Pileup
+  edm::Handle<double> rho;
+  event.getByToken(rho_,rho);
+  
+  // Beam spot
+  edm::Handle<reco::BeamSpot> spot;
+  event.getByToken(beamSpot_,spot);
```

Comment:
It is not clear why ECAL clusters (that supposedly are needed for both KF and PF tracks) are read from `Event` in `produce()`, while these two are read here.